syntax = "proto3";
package users;

service Users {
    rpc GetUsers(GetUsersRequest) returns (GetUsersResponse) {}
    rpc GetUserById(GetUserByIdRequest) returns (GetUserByIdResponse) {}
    rpc CreateUser(CreateUserRequest) returns (CreateUserResponse) {}
    rpc UpdateUser(UpdateUserRequest) returns (UpdateUserResponse) {}
    rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse) {}
    rpc GetUserStream(GetUserStreamRequest) returns (stream GetUserStreamResponse) {}
    rpc CreateUserStream(stream CreateUserStreamRequest) returns (CreateUserStreamResponse) {}
}

message User {
    string id = 1;
    string name = 2;
    string email = 3;
    string password = 4;
}
message GetUsersRequest {
}
message GetUsersResponse {
    repeated User users = 1;
}
message GetUserStreamRequest {
}
message GetUserStreamResponse {
    User user = 1;
}
message GetUserByIdRequest{
    string id = 1;
}
message GetUserByIdResponse{
    User user = 1;
}
message CreateUserRequest {
    User user = 1;
}
message CreateUserResponse {
    User user = 1;
}
message CreateUserStreamRequest {
    User user = 1;
}
message CreateUserStreamResponse {
    string result = 1;
}
message UpdateUserRequest {
    User user = 1;
}
message UpdateUserResponse {
    string id = 1;
}
message DeleteUserRequest {
    User user = 1;
}
message DeleteUserResponse {
    User user = 1;
}
